bpf:overlay: adjust deliver to cilium_host@ingress

smagnani96 opened the pull request:
Prior to this, we were delivering packets to cilium_net@egress.
Given there's no program attached, and we just want the packet to reach
the to-host program, let's adjust the redirection to the cilium_host@ingress
interface, where the `cil_to_host` program is attached (HostFW, etc.).

julianwiedmann commented:
Could you please scroll through the git history on this for a bit, to see if there's a reason why this redirect was chosen (or whether it changed over time) ? I'm wondering whether we **intentionally** chose to perform a veth traversal for access to some processing in that path (netfilter, xfrm, ...) ?

Otherwise this actually looks like a good fix for what's described in https://github.com/cilium/cilium/issues/36329.

julianwiedmann commented:
Note this will need a rebase (and then probably looks a bit simpler too)

smagnani96 committed:
bpf:overlay: refactor ipv6_host_delivery helper

Simple refactor code to align the IPv6 path to the IPv4 ipv4_host_delivery.
This moves some code into a specific new helper `ipv4_host_delivery`.
No functional changes.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:overlay: adjust deliver to cilium_host@ingress

Prior to this, we were delivering host-related packets to cilium_net@egress
after decap in the from-overlay hook. Given there's no program attached there
anymore, and we just want the packet to reach the to-host program, let's
adjust the redirection to the cilium_host@ingress interface,
where the `cil_to_host` program is attached. The to-host program will
take care of the ingress policies (HostFW) before delivering the packet
to the host.

Historically (see [1]), the redirection was to cilium_net@egress because
the to-host program was attached there. Later (see [2]), the to-host program
was moved to cilium_host@ingress, but the redirection was not updated accordingly.
I believe there is no reason nowadays to keep the redirection to cilium_net@egress,
given there's no program/logic there anymore.

In old kernels, a bpf_redirect() with BPF_F_INGRESS would scrub the packet,
as well as a veth traversal. In new kernels, veth traversal or bpf_redirect() in
the same netns would preserve the skb metadata/mark. For this reason,
the packet reaching the `cilium_host@ingress` would appear the same as
before the patch. We simply skip one veth traversal (cilium_net -> cilium_host).

[1] https://github.com/cilium/cilium/commit/d4d9158496d5e32c700f50ed674c27bf3c129aa8
[2] https://github.com/cilium/cilium/commit/1c951ee5b65f9d8bd5cc5b886f4e1855baa8ca82

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
